compiler/internal/codegen/codegen_config_marshalers.go

```diff
@@ -0,0 +1,447 @@
+
+	default:
+		panic(fmt.Sprintf("unsupported type for config unmarshalling: %T", t))
+	}
+}
+
+// readStruct generates code to read a struct from the given iterator. The code generated by this function
+// expects a zero value of the struct to be present in the variable `obj`. This code will be written into the
+// group `f` passed in as a argument.
+//
+// The returned type will be the type definition of the struct.
+func (cb *configUnmarshalersBuilder) readStruct(f *Group, struc *schema.Struct) (returnType *Statement) {
+	fieldTypes := make([]Code, len(struc.Fields))
+
+	f.Id("itr").Dot("ReadObjectCB").Call(Func().Params(
+		Id("itr").Op("*").Qual("github.com/json-iterator/go", "Iterator"),
+		Id("field").String(),
+	).Bool().Block(
+		Switch(Id("field")).BlockFunc(func(f *Group) {
+			for i, field := range struc.Fields {
+				jsonName := field.Name
+				for _, tag := range field.Tags {
+					if tag.Key == "json" {
+						if tag.Name != "" {
+							jsonName = tag.Name
+						}
+					}
+				}
+
+				rhs, returnType := cb.readType(field.Typ, Lit(jsonName))
+				f.Case(Lit(jsonName)).Block(Id("obj").Dot(field.Name).Op("=").Add(rhs))
+
+				fieldTypes[i] = Id(field.Name).Add(returnType)
+
+			}
+		}),
+		Return(True()),
+	))
+
+	return Struct(fieldTypes...)
+}
+
+// typeUnmarshalerFunc returns a `f` function which can be used to read the given value of `typ` and the type
+// that function f returns.
+//
+// The returned function will either be an inline function or an identifier for function defined in the package
+// and is expected to comply with the `config.Unmarshaler[T]` type defined in the runtime.
+func (cb *configUnmarshalersBuilder) typeUnmarshalerFunc(typ *schema.Type) (f *Statement, returnType *Statement) {
+	switch t := typ.Typ.(type) {
+	case *schema.Type_Named:
+		name, returnType := cb.typeUnmarshalerName(t.Named.Id)
+
+		if len(t.Named.TypeArguments) == 0 {
+			return Id(name), returnType
+		} else {
+			returnTypes := make([]Code, len(t.Named.TypeArguments))
+			call := CallFunc(func(f *Group) {
+				for i, arg := range t.Named.TypeArguments {
+					funcToCall, returnType := cb.typeUnmarshalerFunc(arg)
+					returnTypes[i] = returnType
+					f.Add(funcToCall)
+				}
+			})
+
+			f := Id(name).Types(returnTypes...).Add(call)
+
+			return f, returnType.Types(returnTypes...)
+		}
+	default:
+		unmarshaler, returnType := cb.readType(typ, nil)
+
+		return Func().Params(
+			Id("itr").Op("*").Qual("github.com/json-iterator/go", "Iterator"),
+			Id("path").Index().String(),
+		).Params(
+			returnType,
+		).Block(
+			Return(unmarshaler),
+		), returnType
+
+	}
+}
+
+// typeUnmarshalerName returns a generated name for the unmarshaler function for the given type and the type that
+// the decl is.
+func (cb *configUnmarshalersBuilder) typeUnmarshalerName(id uint32) (reader string, rtnTyp *Statement) {
+	return fmt.Sprintf("encoreInternalTypeConfigUnmarshaler_%s", cb.md.Decls[id].Name), Id(cb.md.Decls[id].Name)
```

Comment:
In the past we've defined these as `EncoreInternal_`. I think it's fine they're not exported, but might be good to have the underscore (`encoreInternal_`) to stay consistent